samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemConsumer.java

```diff
@@ -0,0 +1,406 @@
+          // this will call KafkaConsumer.seekToBegin/End()
+          kafkaConsumer.seek(tp, startingOffset); // this value should already be the 'upcoming' value
+        }
+      } catch (Exception e) {
+        // all other exceptions - non recoverable
+        LOG.error("Got Exception while seeking to " + startingOffsetString + " for " + tp, e);
+        throw new SamzaException(e);
+      }
+
+      LOG.info("Changing consumer's starting offset for tp = " + tp + " to " + startingOffsetString);
+
+      // add the partition to the proxy
+      proxy.addTopicPartition(topicPartitions2SSP.get(tp), startingOffset);
+    });
+
+    // start the proxy thread
+    if (proxy != null && !proxy.isRunning()) {
+      LOG.info("Starting proxy: " + proxy);
+      proxy.start();
+    }
+  }
+
+  private void setFetchThresholds() {
+    // get the thresholds, and set defaults if not defined.
+    KafkaConfig kafkaConfig = new KafkaConfig(config);
+
+    Option<String> fetchThresholdOption = kafkaConfig.getConsumerFetchThreshold(systemName);
+    long fetchThreshold = FETCH_THRESHOLD;
+    if (fetchThresholdOption.isDefined()) {
+      fetchThreshold = Long.valueOf(fetchThresholdOption.get());
+      LOG.info("fetchThresholdOption is configured. fetchThreshold=" + fetchThreshold);
+    }
+
+    Option<String> fetchThresholdBytesOption = kafkaConfig.getConsumerFetchThresholdBytes(systemName);
+    long fetchThresholdBytes = FETCH_THRESHOLD_BYTES;
+    if (fetchThresholdBytesOption.isDefined()) {
+      fetchThresholdBytes = Long.valueOf(fetchThresholdBytesOption.get());
+      LOG.info("fetchThresholdBytesOption is configured. fetchThresholdBytes=" + fetchThresholdBytes);
+    }
+
+    int numTPs = topicPartitions2SSP.size();
+    assert (numTPs == topicPartitions2Offset.size());
+
+    LOG.info("fetchThresholdBytes = " + fetchThresholdBytes + "; fetchThreshold=" + fetchThreshold);
+    LOG.info("number of topicPartitions " + numTPs);
+
+    if (numTPs > 0) {
+      perPartitionFetchThreshold = fetchThreshold / numTPs;
+      LOG.info("perPartitionFetchThreshold=" + perPartitionFetchThreshold);
+      if (fetchThresholdBytesEnabled) {
+        // currently this feature cannot be enabled, because we do not have the size of the messages available.
+        // messages get double buffered, hence divide by 2
+        perPartitionFetchThresholdBytes = (fetchThresholdBytes / 2) / numTPs;
+        LOG.info("perPartitionFetchThresholdBytes is enabled. perPartitionFetchThresholdBytes="
+            + perPartitionFetchThresholdBytes);
+      }
+    }
+  }
+
+  @Override
+  public void stop() {
+    LOG.info("Stopping Samza kafkaConsumer " + this);
+
+    if (!stopped.compareAndSet(false, true)) {
+      LOG.warn("attempting to stop stopped consumer.");
+      return;
+    }
+
+    // stop the proxy (with 5 minutes timeout)
+    if (proxy != null) {
+      LOG.info("Stopping proxy " + proxy);
+      proxy.stop(TimeUnit.MINUTES.toMillis(5));
```

Comment:
Why this has to wait for 5 mins to stop the proxy thread. Can we lower the stop timeout to reasonably low value?